source/particle/world.cc

```diff
@@ -0,0 +1,889 @@
+
+              try
+                {
+                  const Point<dim> p_unit = this->get_mapping().transform_real_to_unit_cell(old_cell, it->second.get_location());
+                  if (GeometryInfo<dim>::is_inside_unit_cell(p_unit))
+                    {
+                      ++it;
+                      continue;
+                    }
+                }
+              catch (...)
+                {}
+            }
+
+          // The particle is not in its old cell. Look for the new cell.
+          typename parallel::distributed::Triangulation<dim>::active_cell_iterator cell;
+          try
+            {
+              cell = (GridTools::find_active_cell_around_point<> (this->get_mapping(), this->get_triangulation(), it->second.get_location())).first;
+            }
+          catch (GridTools::ExcPointNotFound<dim> &)
+            {
+              // If we can find no cell for this particle it has left the domain due
+              // to an integration error or an open boundary.
+              lost_particles.insert(*it);
+              particles.erase(it++);
+              continue;
+            }
+
+
+          // Reinsert the particle into our domain if we found its cell
+          // Mark it for MPI transfer otherwise
+          if (cell->is_locally_owned())
+            {
+              const types::LevelInd found_cell = std::make_pair(cell->level(),cell->index());
+              moved_particles_cell.insert(std::make_pair(found_cell, it->second));
+            }
+          else
+            moved_particles_domain.insert(std::make_pair(cell->subdomain_id(),it->second));
+
+          particles.erase(it++);
+        }
+
+      // If particles fell out of the mesh, put them back in
+      move_particles_back_in_mesh(lost_particles,
+                                  moved_particles_cell,
+                                  moved_particles_domain);
+
+      // Reinsert all local particles with their cells
+      particles.insert(moved_particles_cell.begin(),moved_particles_cell.end());
+
+      // Swap lost particles between processors if we have more than one process
+      if (dealii::Utilities::MPI::n_mpi_processes(this->get_mpi_communicator()) > 1)
+        send_recv_particles(moved_particles_domain);
+    }
+
+    template <int dim>
+    void
+    World<dim>::move_particles_back_in_mesh(std::multimap<types::LevelInd, Particle<dim> >            &lost_particles,
+                                            std::multimap<types::LevelInd, Particle<dim> >            &moved_particles_cell,
+                                            std::multimap<types::subdomain_id, Particle<dim> >        &moved_particles_domain)
+    {
+      // TODO: fix this to work with arbitrary meshes. Currently periodic boundaries only work for boxes.
+      const GeometryModel::Box<dim> *geometry
+        = dynamic_cast<const GeometryModel::Box<dim>*> (&this->get_geometry_model());
+
+      if (geometry != 0)
```

Comment:
or do you mean to say that for all other geometries, you simply give up on these particles? If so, replace the assertion with a comment to this effect